### ISSUE TITLE
cmd/go: `go vet ./...` does not correctly resolve symlinks for the top directory

### ISSUE THREAD

**Issue by devoid** [golang/go#12556]
What version of Go are you using (go version)?
go version go1.4.2 darwin/amd64

What operating system and processor architecture are you using?
OS X 10.10.5, amd64

What did you do?

```
$ export GOPATH=$HOME/projects/go-workspace
$ cd $GOPATH/src/github.com/devoid/foobar 
$ go vet ./...
$
$ ln -s $GOPATH/src/github.com/devoid/foobar $HOME/projects/foobar
$ cd $HOME/projects/foobar
$ go vet ./...
vet: error walking tree: stat ../go-workspace/src/github.com/devoid/foobar/main.go: no such file or directory
vet: ../go-workspace/src/github.com/devoid/foobar/main.go: open ../go-workspace/src/github.com/devoid/foobar/main.go: no such file or directory
vet: no files checked
exit status 1
```

What did you expect to see?
Many other go commands `go tool cover`, `godoc` and `go` work fine with this sort of symlink setup.

What did you see instead?
I am not entirely sure. The `go vet` command appears to identify the symlink but can't find the files in it?

**Comment by robpike** [comment]
Symlinks are problematic with these tools. I'm not sure it can be made to work in all cases, or that it's worth trying.

We encourage avoiding symlinks in the Go tree.

**Comment by nmiyake** [comment]
I would like to echo that this issue is problematic for our team. Totally fine with this being the default behavior, but the issue is exacerbated by the difficulty of working around it in Go.

Initially, we tried working around this by launching a new Go process that would invoke `go vet` with a path that we resolved ourselves. Simplified example:

```go
wd, _ := os.Getwd() // path may be a symlink
wd, _ = filepath.EvalSymlinks(wd) // this path is always absolute
cmd := exec.Command("go", "vet")
cmd.Dir = wd // set working directory of command to explicitly be absolute path
cmd.Run() // doesn't work -- os.Getwd() in relPaths() still returns symlink path
```

However, this still doesn't work -- the `os.Getwd()` call in the `relPaths` function of `build.go` still returns the symlink path. It appears that this technically doesn't violate any contracts since `os.Getwd()` states: "If the current directory can be reached via multiple paths (due to symbolic links), Getwd may return any one of them.", but is definitely unintuitive (since the path is explicitly set to be absolute).

We're working around this for now by taking advantage of the fact that `os.Getwd()` consults `$PWD` first, so setting `cmd.Env = append([]string{fmt.Sprintf("PWD=%v", wd)}, os.Environ()...)` works causes this to work as expected. However, we know this is not ideal since it ties the behavior to an undocumented implementation detail that could change at any time.

So, long story short, I second that it would be nice to fix this issue -- although I agree and understand that symlinks into the Go path tree should be avoided, it would be nice to have the option to at least have a plausible work-around to handle this for tools that want to offer a more user-friendly experience.
